--- src/common/clib-package.c.orig	2025-06-18 03:18:07.876000000 +0500
+++ src/common/clib-package.c	2025-06-18 06:06:09.940000000 +0500
@@ -661,11 +661,13 @@
 #else
       res = http_get(json_url);
 #endif
-      json = res->data;
-      _debug("status: %d", res->status);
       if (!res || !res->ok) {
+        http_get_free(res);
+	res = NULL;
         goto download;
       }
+      json = res->data;
+      _debug("status: %d", res->status);
       log = "fetch";
     }
   }
@@ -1489,7 +1491,7 @@
       while (--i >= 0) {
         fetch_package_file_thread_data_t *data = fetchs[i];
         int *status = 0;
-        pthread_join(data->thread, (void **)status);
+        pthread_join(data->thread, (void **)&status);
         free(data);
         fetchs[i] = NULL;
 
@@ -1519,7 +1521,7 @@
     fetch_package_file_thread_data_t *data = fetchs[i];
     int *status = 0;
 
-    pthread_join(data->thread, (void **)status);
+    pthread_join(data->thread, (void **)&status);
 
     (void)pending--;
     free(data);
@@ -1598,6 +1600,7 @@
   }
   fetchs = NULL;
 #endif
+  clib_package_cleanup();
   return rc;
 }
 
@@ -1695,5 +1698,9 @@
 
     hash_free(visited_packages);
     visited_packages = 0;
+    if (clib_package_curl_share) {
+      curl_share_cleanup(clib_package_curl_share);
+      clib_package_curl_share = NULL;
+    }
   }
 }